src/awkward/operations/ak_merge_union_of_records.py

```diff
@@ -41,32 +41,102 @@ def _impl(array, axis, highlevel, behavior):
+                if content.is_option:
+                    tag_for_missing = i
+                    break
+
+            # If at least one recordarray doesn't have this field, we add
+            # a special option
+            if len(field_contents) < len(contents):
+                # Make the tagged content an option, growing by one to ensure we
+                # have a known `None` value to index into
+                tagged_content = field_contents[tag_for_missing]
+                indexedoption_index = backend.index_nplike.arange(
+                    tagged_content.length + 1, dtype=np.int64
+                )
+                indexedoption_index[
+                    index_nplike.shape_item_as_index(tagged_content.length)
+                ] = -1
+                field_contents[
+                    tag_for_missing
+                ] = ak.contents.IndexedOptionArray.simplified(
+                    ak.index.Index64(indexedoption_index), tagged_content
                 )
+
+            # Index of None values in tagged content (content with extra None item at end)
+            index_missing = index_nplike.shape_item_as_index(
+                field_contents[tag_for_missing].length - 1
             )
-        elif layout.is_indexed and layout.content.is_record:
-            record = layout.content
-            # Transpose index-of-record to record-of-index
-            return ak.contents.RecordArray(
-                [
-                    ak.contents.IndexedArray.simplified(
-                        layout.index, c, parameters=layout._parameters
-                    )
-                    for c in record.contents
-                ],
-                record.fields,
-                record.length,
-                backend=backend,
+            # Now build contents for union, by looping over outermost index
+            # Overwrite tags to adjust for new contents length
+            # and use the tagged content for any missing values
+            k = 0
+            for j, content in enumerate(contents):
+                tag_is_j = field_tags == j
+
+                if content.has_field(field):
+                    # Rewrite tags to account for missing fields
+                    field_tags[tag_is_j] = k
+                    k += 1
+
+                else:
+                    # Rewrite tags to point to option content
+                    field_tags[tag_is_j] = tag_for_missing
+                    # Point each value to missing value
+                    field_index[tag_is_j] = index_missing
+
+            outer_field_contents.append(
+                ak.contents.UnionArray.simplified(
+                    ak.index.Index8(field_tags),
+                    ak.index.Index64(field_index),
+                    field_contents,
+                )
             )
-        else:
-            raise ak._errors.wrap_error(TypeError(layout))
+        return ak.contents.RecordArray(
+            outer_field_contents, all_fields, backend=backend
+        )
```

Comment:
This code has not really changed, I only eliminated the index inversion code (this is now handled separately in the new code)